src/elm/Data/DataSource.elm

```diff
@@ -0,0 +1,102 @@
+module Data.DataSource
+    exposing
+        ( DataSourceId(..)
+        , DataSource
+        , decode
+        , decodeList
+        , encodeId
+        )
+
+import Json.Decode as Decode exposing (Decoder)
+import Json.Encode
+
+
+type DataSourceId
+    = DataSourceId Int
+
+
+type Description
+    = NoDescription
+    | Description String
+
+
+type alias DataSource =
+    { id : DataSourceId
+    , title : String
+    , description : Description
+    }
+
+
+decodeList : Decoder (List DataSource)
+decodeList =
+    Decode.list decode
+
+
+decode : Decoder DataSource
+decode =
+    Decode.map3 DataSource id title description
+        |> Decode.map processAbbreviation
```

Comment:
Thoughts on composing here?

```elm
Decode.map3 (processAbbreviation << DataSource) id title description
```

Although perhaps what you really want is a custom constructor function?